### ISSUE TITLE
[Feature]: add apiEndpoint to playwright.config.ts (similar to baseUrl)

### ISSUE THREAD

**Issue by roydekleijn** [microsoft/playwright#23738]
Let us know what functionality you'd like to see in Playwright and what your use case is.
Do you think others might benefit from this as well? **Yes**

From time to time, we like to test our API as well with Playwright. Sometimes the format is different from the baseUrl, that's why an apiEndpoint might be beneficial. Of course, there are some workarounds, but they are not as neath as how baseUrl works. (and apiEndpoint should work)

My example: (is configured as baseUrl)
Web application URL: https://practicrsoftwaretesting.com

API endpoint: https://api.practicesoftwaretesting.com

The various API request function should try to get the value of apiEndpoint first and then append what is left.. ( similar to `.goto()`)

Responds with a thumbs up if you like this addition.

**Comment by yury-s** [comment]
Just to clarify, specifying baseURL does not work for you because you want another base URL for API requests than for page navigations?

```js
import { test, expect } from '@playwright/test';

test.use({
  baseURL: 'https://playwright.dev/'
})

test('test', async ({ request }) => {
  const response = await request.get('/docs/intro')
  expect(response).toBeOK();
})
```

**Comment by roydekleijn** [comment]
That's correct! it would be very neath if we can set it in playwright.config.ts

also, I case we like to combine page navigations with API requests.

**Comment by ltsuda** [comment]
Not exactly what you asked but do you think any of this would be an workaround?

- using a environment variable
- using a constant but you would have to import it when you want it
- something similar to this issue, using fixtures to create an apiBaseURL
  - https://github.com/microsoft/playwright/discussions/23699#discussioncomment-6170857

**Comment by roydekleijn** [comment]
yes, I tried them all..

but it would be a very nice solution if this can be part of `playwright.config.ts` (I think)

**Comment by yury-s** [comment]
As a workaround you can create an [option fixture ](https://playwright.dev/docs/test-fixtures#fixtures-options)`apiBaseURL` and assign that to a value different from `baseURL`, here is a sample project that does it:

https://github.com/yury-s/bug-23738

**Comment by roydekleijn** [comment]
Thanks @yury-s is the workaround the way to go? or would this 'extra' property be part of the default playwright config?

**Comment by yury-s** [comment]
No immediate plans to implement it. If the issue gets enough upvotes we may add it to the default config in the future.

**Comment by deonxxx** [comment]
It would be awesome to get it, but now I’m using external fixture with additional config.json where I put the port. Can’t wait to see it possible in default config! :)

Wysyłane z aplikacji Outlook dla systemu iOS<https://aka.ms/o0ukef>
________________________________
Od: Yury Semikhatsky ***@***.***>
Wysłane: Wednesday, August 23, 2023 10:10:28 PM
Do: microsoft/playwright ***@***.***>
DW: Gedeon Kucharski ***@***.***>; Manual ***@***.***>
Temat: Re: [microsoft/playwright] [Feature]: add apiEndpoint to playwright.config.ts (similar to baseUrl) (Issue #23738)


No immediate plans to implement it. If the issue gets enough upvotes we may add it to the default config in the future.

**Comment by trevorrl** [comment]
Just chiming in to say that I would also love to have an `apiURL` option in the base playwright config instead of needing to use workarounds. Thanks!

**Comment by brunojbarros** [comment]
> As a workaround you can create an [option fixture ](https://playwright.dev/docs/test-fixtures#fixtures-options)`apiBaseURL` and assign that to a value different from `baseURL`, here is a sample project that does it:
> 
> https://github.com/yury-s/bug-23738

I am getting the following error but I don't know what I am missing:


```
No overload matches this call.
  Overload 1 of 6, '(config: PlaywrightTestConfig<APIRequestOptions, {}>): PlaywrightTestConfig<APIRequestOptions, {}>', gave the following error.
    Object literal may only specify known properties, and 'apiInternalURL' does not exist in type 'UseOptions<PlaywrightTestOptions & CustomProperties<APIRequestOptions>, PlaywrightWorkerOptions & CustomProperties<...>>'.
  Overload 2 of 6, '(config: PlaywrightTestConfig<APIRequestOptions, {}>, ...configs: PlaywrightTestConfig<{}, {}>[]): PlaywrightTestConfig<APIRequestOptions, {}>', gave the following error.
    Object literal may only specify known properties, and 'apiInternalURL' does not exist in type 'UseOptions<PlaywrightTestOptions & CustomProperties<APIRequestOptions>, PlaywrightWorkerOptions & CustomProperties<...>>'.ts(2769)
test.d.ts(1437, 3): The expected type comes from property 'use' which is declared here on type 'PlaywrightTestConfig<APIRequestOptions, {}>'
test.d.ts(1437, 3): The expected type comes from property 'use' which is declared here on type 'PlaywrightTestConfig<APIRequestOptions, {}>'
```

![image](https://github.com/microsoft/playwright/assets/12484408/a85aea65-24e4-4777-8c33-481e68227a83)

I create this repo:
https://github.com/brunojbarros/playwright-overload-error-23738



In my case, I need more than only two base urls, so maybe it could be a useful a `host:string[] | undefined` instead of `baseURL: string | undefined`.

**Comment by brunojbarros** [comment]
In my case, I need more than only two base urls, so maybe it could be a useful a `host:string[] | undefined` instead of `baseURL: string | undefined`.

**Comment by yurii-sav** [comment]
Voting for `apiUrl`, and thanks for the workaround!

**Comment by sdt-jchau** [comment]
Another vote for apiUrl.  It's pretty common for the baseUrl to be different than apiUrl.

**Comment by StevenBoutcher** [comment]
Playwright can't claim to support API testing as a first-class citizen if it doesn't have a base URL for the API in the config. It makes it feel like API testing is just an "add-on" capability that doesn't receive the same attention as browser testing. API tests should be divorced from the browser to be truly API-level, anyway. Not having a feature like this contributes in a small way to anti-patterns like using the `page` fixture for API tests.

**Comment by nsnogren-trend** [comment]
Voted for apiUrl. It's a shame having to either split URL configuration between 2 places, or choose not to use Playwright's built in baseUrl.

My solution:

```
// https://github.com/microsoft/playwright/issues/23738
declare module '@playwright/test' {
  interface PlaywrightTestOptions {
    apiBaseURL?: string;
  }
}
```

```
projects: [
    {
      name: 'local',
      use: { 
        ...devices['Desktop Chrome'],
        baseURL: 'http://localhost:4200',
        apiBaseURL: 'http://localhost:3000',
      },
    },
```

```
const apiBaseUrl = testInfo.project.use.apiBaseURL;
    if (!apiBaseUrl) {
      throw new Error(`API base URL not configured for project "${testInfo.project.name}". Add apiBaseURL to project use config in playwright.config.ts`);
    }
```

**Comment by testsmith-io** [comment]
@yury-s or @debs-obrien quite some votes... can this be implemented?
